src/Pos/Block/Logic.hs

```diff
@@ -773,39 +779,60 @@ createMainBlockFinish
-    let prependToUndo undos tx =
-            fromMaybe (error "Undo for tx not found")
-                      (HM.lookup (fst tx) txUndo) : undos
-    lift $ inAssertMode $ verifyBlocksPrefix (one (Right blk)) >>= \case
-        Left err ->
-            assertionFailed $ sformat ("We've created bad block: "%stext) err
-        Right _ -> pass
-    (pModifier, verUndo) <-
-        runExceptT (usVerifyBlocks False (one (toUpdateBlock (Right blk)))) >>= \case
-            Left _ ->
-                throwError "Couldn't get pModifier while creating MainBlock"
-            Right o -> pure o
-    let blockUndo = Undo (reverse $ foldl' prependToUndo [] sortedTxs)
-                         pskUndo
-                         (verUndo ^. _Wrapped . _neHead)
-    evaluateNF_ (blockUndo, blk)
+    unchecked@(uncheckedBlock, _, _) <- createBlundFromMemPool
+    (block, undo, pModifier) <-
+        verifyCreatedBlock uncheckedBlock >>=
+        either fallbackCreateBlock (const $ pure unchecked)
     logDebug "Created main block/undos, applying"
-    lift $ blk <$ applyBlocksUnsafe getPeers (one (Right blk, blockUndo)) (Just pModifier)
+    lift $ block <$ applyBlocksUnsafe getPeers (one (Right block, undo)) (Just pModifier)
   where
+    createBlundFromMemPool :: ExceptT Text m (MainBlock ssc, Undo, PollModifier)
+    createBlundFromMemPool = do
+        -- Get MemPool
+        (localTxs, txUndo) <- getLocalTxsNUndo
+        sscData <- sscGetLocalPayload @ssc slotId
+        usPayload <- note onNoUS =<< lift (usPreparePayload slotId)
+        (localPSKs, pskUndo) <- lift getProxyMempool
+        -- Create block
+        let convertTx (txId, (tx, _, _)) = WithHash tx txId
+        sortedTxs <- maybe onBrokenTopo pure $ topsortTxs convertTx localTxs
+        sk <- npSecretKey . ncNodeParams <$> getNodeContext
+        -- for now let's be cautious and not generate blocks that are larger than
+        -- maxBlockSize/4
+        sizeLimit <- fromIntegral . toBytes . (`div` 4) <$> UDB.getMaxBlockSize
+        block <- createMainBlockPure
+                     sizeLimit prevHeader sortedTxs pSk
+                     slotId localPSKs sscData usPayload sk
+        -- Create undo
+        (pModifier, verUndo) <-
+            runExceptT (usVerifyBlocks False (one (toUpdateBlock (Right block)))) >>=
+            either (const $ throwError "Couldn't get pModifier while creating MainBlock") pure
+        let undo = Undo (reverse $ foldl' (prependToUndo txUndo) [] sortedTxs)
+                        pskUndo
+                        (verUndo ^. _Wrapped . _neHead)
+        () <- (undo `deepseq` block) `deepseq` pure ()
+        pure (block, undo, pModifier)
     onBrokenTopo = throwError "Topology of local transactions is broken!"
     onNoUS = "can't obtain US payload to create block"
+    prependToUndo txUndo undos tx =
+        fromMaybe (error "Undo for tx not found") (HM.lookup (fst tx) txUndo) : undos
+    verifyCreatedBlock block = lift $
```

Comment:
I mean, it should be written in code, because it takes a big amount of time to infer type of it.